core/domain/user_services.py

```diff
@@ -2287,3 +2291,267 @@ def get_dashboard_stats(user_id):
+
+        # Get the furthest reached checkpoint in current exploration.
+        furthest_reached_checkpoint_in_current_exploration = (
+            _get_most_distant_reached_checkpoint_in_current_exploration(
+                checkpoints_in_current_exploration,
+                checkpoints_in_furthest_reached_exploration,
+                exploration_user_model.furthest_reached_checkpoint_state_name
+            )
+        )
+
+        # If the furthest reached checkpoint doesn't exist in current
+        # exploration.
+        if furthest_reached_checkpoint_in_current_exploration is None:
+            exploration_user_model.furthest_reached_checkpoint_exp_version = (
+                exp_version)
+            exploration_user_model.furthest_reached_checkpoint_state_name = (
+                state_name)
+        else:
+            # If furthest reached checkpoint is behind most recently
+            # reached checkpoint.
+            if (checkpoints_in_current_exploration.index(
+                furthest_reached_checkpoint_in_current_exploration) <= (
+                    checkpoints_in_current_exploration.index(state_name))):
+                (
+                    exploration_user_model
+                    .furthest_reached_checkpoint_exp_version
+                ) = exp_version
+                (
+                    exploration_user_model
+                    .furthest_reached_checkpoint_state_name
+                ) = state_name
+
+    exploration_user_model.most_recently_reached_checkpoint_exp_version = (
+        exp_version)
+    exploration_user_model.most_recently_reached_checkpoint_state_name = (
+        state_name)
+    exploration_user_model.update_timestamps()
+    exploration_user_model.put()
+
+
+def set_user_has_viewed_lesson_info_modal_once(user_id):
+    """Updates the user's settings once he has viewed the lesson info modal.
+
+    Args:
+        user_id: str. The Id of the user.
+    """
+
+    user_settings = get_user_settings(user_id)
+    user_settings.mark_lesson_info_modal_viewed()
+    _save_user_settings(user_settings)
+
+
+def update_learner_checkpoint_progress_on_restart(user_id, exploration_id):
+    """Sets the most recently reached checkpoint on restart exploration event.
+
+    Args:
+        user_id: str. The Id of the user.
+        exploration_id: str. The Id of the exploration.
+    """
+    exploration_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exploration_id)
+    if exploration_user_model is None:
+        exploration_user_model = user_models.ExplorationUserDataModel.create(
+            user_id, exploration_id)
+    exploration_user_model.most_recently_reached_checkpoint_exp_version = None
+    exploration_user_model.most_recently_reached_checkpoint_state_name = None
+    exploration_user_model.update_timestamps()
+    exploration_user_model.put()
+
+
+def synchronize_learner_checkpoint_progress_with_current_exp_version(
```

Comment:
```suggestion
def sync_learner_checkpoint_progress_with_current_exp_version(
```

Reply:
Just make this change to shorten function name a bit

Reply:
Done!